### DISCUSSION TITLE
Can we force `__typename` to be always present and not to be `undefined`?

### DISCUSSION THREAD

**Discussion by vinassefranche** [0no-co/gql.tada#423] [Q&A] [answered]
Hi there!

Using apollo client, the `__typename` is always fetched so we know it will always be there. We use it intensively as a typeguard in our code base and it's added on everything we fetch with our codegen.
Here, if I'm not mistaken, I can see that it's not added on fragments and it's added at the top level in queries typed as `T | undefined` .
Is there a way to configure gql.tada to always add `__typename` and without the `| undefined` part? If not, would it be something that could be added?

Thanks for your answer!

**Comment by kitten** [comment, marked as answer]
This is explicitly not supported and `__typename` is only included when fetching. Generally, we deem this more correct and an easily spottable error, since using it will immediately result in TS type errors.

In urql Graphcache, we actually adhere to `__typename` inclusion, and Graphcache will reselect each selection, omitting `__typename` if it isn't included.

The problem here is that while Apollo Client implicitly queries it, it's still an implicit selection that could be removed, i.e. if you look at what Graphcache does, it's technically more "correct" behaviour to adhere to the original selection, which is why we can't assume the field to be there. (A link also exists for Apollo afaik to actually remove the `__typename` property explicitly from responses if it's omitted in the selection set)

We don't really aim to introduce options for all possible implicit behaviour a GraphQL client may introduce, and instead opt to provide a correct and predictable default instead.

**Comment by vinassefranche** [comment]
Alright. It might limit our adoption of gql.tada as it would require to update all our fragments and queries (even if it can be made little by little) but thanks for the detailed answer. I understand the point.
